src/protocol.rs

```diff
@@ -0,0 +1,760 @@
+#![allow(dead_code)]
+
+use mockall_double::double;
+
+#[double]
+use crate::targets::esp32c3 as target;
+
+pub trait InputIO {
+    fn recv(&mut self) -> u8;
+    fn send(&mut self, data: &[u8]);
+}
+
+type FlashFunc = fn(addr: u32, data: *const u8, len: u32) -> Result<(), Error>;
+
+use target::*;
+use slip::*;
+use core::mem::size_of;
+use core::slice;
+use core::cmp::min;
+use crate::commands::*;
+use crate::commands::CommandCode::*;
+use crate::commands::Error::*;
+use md5::{Md5, Digest};
+use crate::miniz_types::*;
+use crate::dprintln;
+
+const DATA_CMD_SIZE: usize = size_of::<DataCommand>();
+const CMD_BASE_SIZE: usize = size_of::<CommandBase>();
+
+const FLASH_SECTOR_SIZE: u32 = 4096;
+const MAX_WRITE_BLOCK: u32 = 0x4000;
+pub struct Stub<'a> {
+    io: &'a mut (dyn InputIO + 'a),
```

Comment:
For performance reasons, we may wish to remove `dyn` and instead have a run-time selectable enum with all possible io options. Lets keep this like it is for now though.

Reply:
My understanding was that `dyn` is equivalent of virtual call in C++, and thus is quite inexpensive. Or Am I missing something?

Reply:
Yes, it's the same, but as with C++ virtual calls, there is some overhead dispatching via the virtual method table. In our case we will know every single IO method at compile time, we just need to be able to switch between them. In this case, an enum would provide us with direct call access to the chosen IO without having to go through the virtual method table. More detail in the README for this crate: https://crates.io/crates/enum_dispatch.

Don't worry for now, let's keep it as `dyn`.